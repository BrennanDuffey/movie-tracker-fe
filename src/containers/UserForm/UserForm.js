import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from "react-router-dom";
import { loginUser, isLoading, errorMessage, setFavorites } from '../../actions';
import { agnosticFetch } from '../../utils/apiCalls/apiCalls'


export class UserForm extends Component {
  constructor() {
    super()
    this.state= {
      name: '',
      email: '',
      password: '',
      successfulLogin: false
    }
  }

  handleChange = (e) => {
    let {name, value} = e.target;
    this.setState({[name]: value});
  }

  fetchFavorites = async (id) => {
    const url = `http://localhost:3000/api/users/${id}/favorites`
    try {
      const response = await agnosticFetch(url, null)
      const favorites = await response.data
      this.props.setFavorites(favorites)
    } catch {
      this.props.setErrorMessage('Error in fetching user favorites')
    }
  }

  handleLogin = (e) =>{
    e.preventDefault();
    const url = 'http://localhost:3000/api/users';
    const init = this.createInit(this.state)
    agnosticFetch(url, init)
      .then(result => {
        this.setState({successfulLogin: true})
        this.props.loginUser(result.data);
        this.fetchFavorites(result.data.id)
      })
      .catch(error => {
        this.props.setErrorMessage('Incorrect email/password combination')
        window.alert('Incorrect email/password combination')
      });
  };

  handleSignup = async (e) => {
    e.preventDefault();
    this.props.isLoading(true);
    const url = 'http://localhost:3000/api/users/new';
    const init = this.createInit(this.state)
    try {
      const result = await agnosticFetch(url, init);
      const newUser = { 
        id: result.id, 
        name: this.state.name, 
        email: this.state.email, 
        password: this.state.password 
      };
      this.props.loginUser(newUser)
      this.setState({successfulLogin: true})
    } catch (error) {
      this.props.setErrorMessage('Email already used for an account')
      window.alert('Email already used for an account')
    }
    this.props.isLoading(false)
  };

  createInit(body) {
    const init = {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify(body)
    }
    return init
  };

  render() {
    if (this.state.successfulLogin) {
      return <Redirect to='/' />
    }

    return (
      <section className="UserForm">
        <form className="existingUser" onSubmit={this.handleLogin}>
          <h2>Existing User</h2>
          <span className="EU-input">
            <input
              name="email"
              onChange={this.handleChange}
              type="email"
              placeholder="Email@email.com"
            />
            <input
              name="password"
              onChange={this.handleChange}
              type="password"
              placeholder="Password"
            />
          </span>
          <input className="user-submit" type="submit" value="Login" />
        </form>
        <div className="divider" />
        <form className="newUser" onSubmit={this.handleSignup}>
          <h2>New User: Create Account</h2>
          <span className="EU-input">
            <input
              name="name"
              maxLength="10"
              onChange={this.handleChange}
              type="text"
              placeholder="Name"
            />
            <input
              name="email"
              onChange={this.handleChange}
              type="email"
              placeholder="Email@email.com"
            />
            <input
              name="password"
              onChange={this.handleChange}
              type="password"
              placeholder="Password"
            />
          </span>
          <input className="user-submit" type="submit" value="Sign-Up" />
        </form>
      </section>
    );
  }
}

UserForm.propTypes={
  history: PropTypes.object,
  isLoading: PropTypes.func,
  location: PropTypes.object,
  loginUser: PropTypes.func,
  match: PropTypes.object,
  setErrorMessage: PropTypes.func,
  setFavorites:PropTypes.func
}

const mapDispatchToProps = (dispatch) => ({
  loginUser: (user) => dispatch(loginUser(user)),
  isLoading: (bool) => dispatch(isLoading(bool)),
  setErrorMessage: (message) => dispatch(errorMessage(message)),
  setFavorites: (favorites) => dispatch(setFavorites(favorites))
});

export default connect(null, mapDispatchToProps)(UserForm)