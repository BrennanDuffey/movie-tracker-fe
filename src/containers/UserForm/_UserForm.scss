.UserForm {
  background-color: black;
  width: 89vw;
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
  background: url('https://images.unsplash.com/photo-1489599849927-2ee91cede3ba?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80');
  background-size: cover;

    h2 {
      text-align: center;
      color: white;
      text-shadow: 2px 2px 2px #500000;
    }

    input {
      margin: 20px 0;
      border-radius: 15px;
      height: 30px;
      padding: 5px 15px 0 15px;
      background: #500000;
      color: white;
    }
    input::placeholder {
      color: silver;
    }
    .user-submit {
      height: 50px;
      width: 160px;
      margin: 20px auto;
    }
}

.divider {
  content: '';
  position: absolute;
  left: 49.78%;
  width: 5px;
  height: 45%;
  border-radius: 30%;
  background-color: #500000;
}

.existingUser,
.newUser {
  display: grid;
  grid-template-rows: 1fr 3fr 1fr;
  grid-template-columns: 100%;
  justify-content: center;
  align-items: center;
  background-color: #970000;
  height: 50vh;
  width: 25vw;
  padding: 20px;
  box-shadow: black 5px 5px 10px;

    input {
      background: #3a3a3a;
      width: 70%;
      }

    .user-submit {
      background: #500000;
      text-transform: uppercase;

        &:hover {
          background: #3a3a3a;
          box-shadow: 2px 2px 6px white;
          cursor: pointer;
        }
    }

    .EU-input {
      display: flex;
      justify-content: center;
      flex-direction: column;
      align-items: center;
    }
}

.existingUser {
  border-radius: 35px 0 0 35px;
  margin: 0 0 0 20px;
}

.newUser {
  border-radius: 0 35px 35px 0;
  margin: 0 20px 0 0;
}