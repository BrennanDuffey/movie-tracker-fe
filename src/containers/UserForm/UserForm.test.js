import {UserForm} from './UserForm';
import React from 'react';
import {shallow} from 'enzyme';

describe('UserForm', ()=>{
  let wrapper;

  beforeEach(()=>{
    wrapper= shallow(<UserForm />);
  })

  it('should match snapshot', ()=>{
    expect(wrapper).toMatchSnapshot();
  });
})