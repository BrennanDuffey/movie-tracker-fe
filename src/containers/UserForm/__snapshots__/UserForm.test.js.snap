// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm should match snapshot 1`] = `
<section
  className="UserForm"
>
  <form
    className="existingUser"
    onSubmit={[Function]}
  >
    <h2>
      Existing User
    </h2>
    <span
      className="EU-input"
    >
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email@email.com"
        type="email"
      />
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
    </span>
    <input
      className="user-submit"
      type="submit"
      value="Login"
    />
  </form>
  <div
    className="divider"
  />
  <form
    className="newUser"
    onSubmit={[Function]}
  >
    <h2>
      New User: Create Account
    </h2>
    <span
      className="EU-input"
    >
      <input
        maxLength="10"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        type="text"
      />
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email@email.com"
        type="email"
      />
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
    </span>
    <input
      className="user-submit"
      type="submit"
      value="Sign-Up"
    />
  </form>
</section>
`;
